staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,2118 @@
+                }
+                _ => {
+                    trace!("Ctrl queue: msg type 0x{:?} unknown", request.msg_type);
+                    Err(Error::HandleEventUnknown)
+                }
+            }?;
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1));
+            }
+
+            desc_chain
+                .memory()
+                .write_slice(response.as_slice(), desc_response.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), desc_response.len())
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn process_tx_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_tx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty()");
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.len() != 2 {
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if (desc_request.len() as usize) > size_of::<VirtioCanFrame>()
+                || (desc_request.len() as usize) < (size_of::<VirtioCanFrame>() - 64)
+            {
+                trace!(
+                    "desc_len: {}, virtio_can_len: {:?}\n",
+                    desc_request.len() as usize,
+                    size_of::<VirtioCanFrame>()
+                );
+                return Err(Error::UnexpectedTxDescriptorSize(
+                    size_of::<VirtioCanFrame>() - 64,
+                    size_of::<VirtioCanFrame>(),
+                    desc_request.len(),
+                ));
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioCanFrame>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1));
+            }
+
+            let response = if self.controller.read().unwrap().ctrl_state == CAN_CS_STOPPED {
+                trace!("Device is stopped!");
+                VIRTIO_CAN_RESULT_NOT_OK
+            } else {
+                let _response = match self.check_tx_frame(request) {
+                    Ok(frame) => {
+                        // If the VIRTIO_CAN_F_CAN_LATE_TX_ACK is negotiated sent the
```

Comment:
I think the spec is not very clear here. It states:

> VIRTIO_CAN_F_LATE_TX_ACK (3) The virtio CAN device marks transmission requests from the Txq as
used after the CAN message has been transmitted on the CAN bus. If this feature bit has not been ne-
gotiated, the device is allowed to mark transmission requests already as used when the CAN message
has been scheduled for transmission but might not yet have been transmitted on the CAN bus.

If I understand correctly, when this feature has *not* been negotiated, the device may mark the transmission request as used before the message has been sent through the CAN bus. In this case, the used notification would be synchronous with the reception of the request in tx. Conversely, if we mark the request as used *after* the message has been transmitted, the used notification would be asynchronously with the reception of the request in tx. Does it make sense? Current implementation does not negotiate this feature so either semantics are allowed. I think the spec should use **MUST** in case the feature is negociated and **MAY** in case is not.